pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java

```diff
@@ -163,32 +196,98 @@ public Boolean hasTrackingScan(String kitLabel) {
+                stmt.setString(1, kitRequestShipping.getDdpLabel());
+                ResultSet rs = stmt.executeQuery();
+                if (!rs.next()) {
+                    String errorMessage =
+                            "No rowcount when querying kits with label " + kitRequestShipping.getDdpLabel();
+                    dbVals.resultValue = new ScanError(kitRequestShipping.getDdpLabel(), errorMessage);
+                    return dbVals;
+                }
+                int numKitsWithLabel = rs.getInt(1);
+
+                if (numKitsWithLabel == 0) {
+                    dbVals.resultValue = new ScanError(kitRequestShipping.getDdpLabel(),
+                            "kit " + kitRequestShipping.getDdpLabel() + " not found.");
+                    return dbVals;
+                } else if (numKitsWithLabel > 1) {
+                    // todo arz is this an error?  Expected to have multiple kits with the same label?
+                    dbVals.resultValue = new ScanError(kitRequestShipping.getDdpLabel(),
+                            "Found " + numKitsWithLabel + " kits with label " + kitRequestShipping.getDdpLabel());
+                    return dbVals;
+                } // else 1 kit, which is what we expect
+            } catch (SQLException ex) {
+                logger.error("Not able to query kits for ddpLabel " + kitRequestShipping.getDdpLabel(), ex);
+                dbVals.resultValue = new ScanError(kitRequestShipping.getDdpLabel(),
+                        "An unexpected error has occurred");
+            }
+
+            try (PreparedStatement stmt = conn.prepareStatement(SET_DDP_KIT_SCAN_INFO_BY_DDP_LABEL_IF_NOT_SET_ALREADY)) {
                 stmt.setLong(1, System.currentTimeMillis());
                 stmt.setString(2, userId);
                 stmt.setString(3, kitRequestShipping.getKitLabel());
                 stmt.setString(4, kitRequestShipping.getDdpLabel());
+                stmt.setString(5, kitRequestShipping.getKitLabel());
                 int rowsAffected = stmt.executeUpdate();
-                if (rowsAffected != 1) {
-                    dbVals.resultValue = new ScanError(kitRequestShipping.getDdpLabel(), "ddp_label "
-                            + kitRequestShipping.getDdpLabel() + " does not exist or already has a Kit Label");
-                } else {
-                    logger.info("Updated kitRequests w/ ddp_label " + kitRequestShipping.getDdpLabel());
-                    if (StringUtils.isNotBlank(kitRequestShipping.getBspCollaboratorParticipantId())) {
+                if (rowsAffected == 1) {
+                    logger.info("Updated ddp_kit.kit_label to {} for kit request  {}.",
+                            kitRequestShipping.getKitLabel(), kitRequestShipping.getDdpLabel());
+                    if (kitRequestShipping.hasBSPCollaboratorParticipantId()) {
+                        // todo arz how does this work?  A scan error with no error?
```

Comment:
@pegahtah can you help me understand this special block of code that shows up when there's a bsp collab participant id?

Reply:
Of course, for understanding this I need to first tell you  that the ScanError is used to communicate anything (not just errors) with the frontend. If the kit has a bsp participant id, we will send it to the frontend. So the frontend will show the successful scan using the returned response. (GP wants to see that as a final step to make sure they are sending the kit to the right participant)
 As an example Angio Dev returned this `[{"kit":"W9O6M23K73T36N8","shortId":"ASCProject_P9846A"}]` for a successful scan, in which `kit` is the `DSM shipping label` or `ddpLabel` and frontend uses it to display the bspParticipantId under the correct corresponding scan line
![Screenshot 2023-11-22 at 11 09 45 AM](https://github.com/broadinstitute/ddp-study-server/assets/47120870/1abdba87-dff7-45f4-a6b9-11fb65c65a2f)